File: src/at_application.rb
require 'quickfix_ruby'

class Application < Quickfix::Application

	def initiatlize
		@orderIDs = Hash.new
	end

	def onCreate(sessionID)
		reset()
		session = Quickfix::Session.lookupSession( sessionID )
		session.reset() if session != nil
	end

	def onLogon(sessionID)
	end

	def onLogout(sessionID)
		reset()
	end

	def toAdmin(sessionID, message)
	end

	def fromAdmin(sessionID, message)
	end

	def toApp(sessionID, message)
	end

	def fromApp(message, sessionID)
		msgType = Quickfix::MsgType.new
		message.getHeader().getField( msgType )
		if( msgType.getValue() == Quickfix::MsgType_NewOrderSingle() )
			echo = Quickfix::Message.new( message )
			possResend = Quickfix::PossResend.new( false )
			if( message.getHeader().isSetField( possResend ) )
				message.getHeader().getField( possResend )
			end

			clOrdID = Quickfix::ClOrdID.new
			message.getField( clOrdID )

			if( possResend.getValue() )
				if( @orderIDs.has_key?([clOrdID.getString(), sessionID.toString()]) )
					return
				end
			end
			@orderIDs[[clOrdID.getString(), sessionID.toString()]] = [clOrdID.getString(), sessionID.toString()]
			Quickfix::Session::sendToTarget( echo, sessionID )
		elsif( msgType.getValue() == Quickfix::MsgType_SecurityDefinition() )
			echo = Quickfix::Message.new( message )
			Quickfix::Session::sendToTarget( echo, sessionID )
		else
			raise Quickfix::UnsupportedMessageType.new
		end
	end

	def reset
		@orderIDs = Hash.new
	end

end

File: src/at_application.py
import quickfix as fix

class Application(fix.Application):
    orderIDs = {}

    def onCreate(self, sessionID):
        self.reset()
        session = fix.Session.lookupSession( sessionID )
        if session == None:
            session.reset()

    def onLogon(self, sessionID): return
    def onLogout(self, sessionID): return
    def toAdmin(self, sessionID, message): return
    def fromAdmin(self, sessionID, message): return
    def toApp(self, sessionID, message): return

    def fromApp(self, message, sessionID):
        msgType = fix.MsgType()
        message.getHeader().getField( msgType )
        if( msgType.getValue() == fix.MsgType_NewOrderSingle ):
            echo = fix.Message( message )
            possResend = fix.PossResend( False )
            if( message.getHeader().isSetField( possResend ) ):
                message.getHeader().getField( possResend )

            clOrdID = fix.ClOrdID()
            message.getField( clOrdID )

            if( possResend.getValue() ):
                if( clOrdID.getString() in self.orderIDs ):
                    return
            self.orderIDs[clOrdID.getString()] = clOrdID.getString()
            fix.Session.sendToTarget( echo, sessionID )
        elif( msgType.getValue() == fix.MsgType_SecurityDefinition ):
            echo = fix.Message( message )
            fix.Session.sendToTarget( echo, sessionID )
        else:
            raise fix.UnsupportedMessageType()

    def reset( self ):
        self.orderIDs = {}

